src/client.toit

```diff
@@ -10,133 +10,244 @@ import .tcp
+        if not closed_ and e and e != CLOSED_ERROR:
+          logger_.error "error processing transport" --tags={"error": e}
+        run_task_ = null
+
+  /**
+  Starts a new transaction.
+  */
+  enter -> int:
+    signal_.wait: transactions_.size < max_concurrent_transactions_
+    identifier := next_identifier_++
+    transactions_.update identifier --if_absent=null: throw "invalid identifier"
+    return identifier
+
+  /**
+  Ends the transaction with the given $identifier.
+
+  If the manager later receives a message for this transaction it will be treated as an error.
+  */
+  leave identifier:
+    transactions_.remove identifier
+    // Allow new transactions to start and error out transactions that are stuck in $get_frame.
+    signal_.raise
+
+  /**
+  Reads a frame for the transaction with the given $identifier.
+  */
+  read identifier -> Frame:
+    response := null
+    signal_.wait:
+      response = transactions_.get identifier --if_absent=: throw TRANSACTION_CLOSED_ERROR
+      response != null
+    if transactions_.contains identifier: transactions_[identifier] = null
+    if response is not Frame: throw response
+    return response
+
+  /**
+  Writes a frame for the transaction with the given $identifier.
+  */
+  write identifier/int frame/Frame -> none:
+    assert: frame.identifier == identifier
+    transport_.write frame
+
+  dispatch_ frame_or_exception -> bool:
+    transaction_identifier := ?
+    if frame_or_exception is Frame:
+      transaction_identifier = frame_or_exception.identifier
+    else if frame_or_exception is InvalidFrameException:
+      transaction_identifier = frame_or_exception.transaction_identifier
+    else:
+      unreachable
+
+    if transaction_identifier == Frame.NO_TRANSACTION_ID:
+      // We are using a framer that doesn't support transactions.
+      // That also means that there must be at most one transaction at a time.
+      if transactions_.size > 1: throw "INVALID_STATE"
+      if transactions_.is_empty:
+        // Nobody is waiting anymore.
+        return false
+      transaction_identifier = transactions_.keys.first
+
+    // There should never be more than one response frame for a given transaction.
+    if (transactions_.get transaction_identifier) != null: return false
+    transactions_.update transaction_identifier --if_absent=(:return false): frame_or_exception
+    signal_.raise
+    return true
+
+  /**
+  Closes the transaction manager.
+
+  If there are existing transactions pending they are aborted with a $CLOSED_ERROR.
+  */
+  close:
+    if closed_: return
+    closed_ = true
+    critical_do:
+      // TODO(florian): should we close the transport even if we didn't create it?
+      transport_.close
+      // Terminate non-completed sessions with an closed error.
+      transactions_.map --in_place: | _ frame |
+        frame ? frame : CLOSED_ERROR
+      signal_.raise
+      if run_task_: run_task_.cancel
+
+  run:
     try:
       while true:
```

Comment:
while not Task.current.canceled?

Probably not necessary because of the return if frame is null, but might be more robust against future changes.

Reply:
Was on the edge, but I can't use `Task.current` as we want the package to work with toit.io as well.
And there we don't have `Task.current` yet. I could just use the deprecated `task` but then we would have yet another health-gold issue.
For now leaving as is, but adding an issue.